src/kvstore/raftex/RaftPart.cpp

```diff
@@ -1470,64 +1429,101 @@ void RaftPart::processAppendLogRequest(const cpp2::AppendLogRequest& req,
-              << ". So need to rollback to last committedLogId_ " << committedLogId_;
-    if (wal_->rollbackToLog(committedLogId_)) {
-      lastLogId_ = wal_->lastLogId();
-      lastLogTerm_ = wal_->lastLogTerm();
-      resp.set_last_log_id(lastLogId_);
-      resp.set_last_log_term(lastLogTerm_);
-      LOG(INFO) << idStr_ << "Rollback succeeded! lastLogId is " << lastLogId_ << ", logLogTerm is "
-                << lastLogTerm_ << ", committedLogId is " << committedLogId_ << ", term is "
-                << term_;
-    }
-    resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
-    return;
+  if (req.get_last_log_id_sent() == lastLogId_ && req.get_last_log_term_sent() == lastLogTerm_) {
+    // nothing to do
+    // just append log later
   } else if (req.get_last_log_id_sent() > lastLogId_) {
     // There is a gap
     LOG(INFO) << idStr_ << "Local is missing logs from id " << lastLogId_ << ". Need to catch up";
     resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
     return;
-  } else if (req.get_last_log_id_sent() < lastLogId_) {
-    // TODO(doodle): This is a potential bug which would cause data not in
-    // consensus. In most case, we would hit this path when leader append logs
-    // to follower and timeout (leader would set lastLogIdSent_ = logIdToSend_ -
-    // 1 in Host). **But follower actually received it successfully**. Which
-    // will explain when leader retry to append these logs, the LOG belows is
-    // printed, and lastLogId_ == req.get_last_log_id_sent() + 1 in the LOG.
-    //
-    // In fact we should always rollback to req.get_last_log_id_sent(), and
-    // append the logs from leader (we can't make promise that the logs in range
-    // [req.get_last_log_id_sent() + 1, lastLogId_] is same with follower).
-    // However, this makes no difference in the above case.
-    LOG(INFO) << idStr_ << "Stale log! Local lastLogId " << lastLogId_ << ", lastLogTerm "
-              << lastLogTerm_ << ", lastLogIdSent " << req.get_last_log_id_sent()
-              << ", lastLogTermSent " << req.get_last_log_term_sent();
-    resp.set_error_code(cpp2::ErrorCode::E_LOG_STALE);
-    return;
+  } else {
+    // check the last log term is matched or not
+    int reqLastLogTerm = wal_->getLogTerm(req.get_last_log_id_sent());
+    if (req.get_last_log_term_sent() != reqLastLogTerm) {
+      LOG(INFO) << idStr_ << "The local log term is " << reqLastLogTerm
+                << ", which is different from the leader's prevLogTerm "
+                << req.get_last_log_term_sent() << ", the prevLogId is "
+                << req.get_last_log_id_sent() << ". So ask leader to send logs from committedLogId "
+                << committedLogId_;
+      TermID committedLogTerm = wal_->getLogTerm(committedLogId_);
+      if (committedLogTerm > 0) {
+        resp.set_last_log_id(committedLogId_);
+        resp.set_last_log_term(committedLogTerm);
+      }
+      resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
+      return;
+    }
```

Comment:
emm...reseting last_log_id directly to committedLogId_ is kind of..brutal, but it's right and work.
A possible optimization when mismatch happends:

From follower side:

1. when Term(leader) < Term(follower), set resp.last_log_id to the index of first occurance of Term(follower) in follower's log entries.
2. when Term(leader) > Term(follower), just decrement last_log_id

apply the similar process to update follower's `nextIndex`(term from raft thesis, think it's similar to our last_log_id) when leader receive a append entries response. 

The above algorithm will make raft quickly find the match log without retransfering entreis already existed in follower's log, but this would further complex raft algorithm(indeed we can still further optimize this algorithm by using a updated version of binary search), so I'm not sure whether it's worth doing so.

Reply:
> reseting last_log_id directly to committedLogId_ is kind of..brutal, but it's right and work.

Totally agreed. I will think more about it. It need some complex process to make the `last_log_id` right, as for your suggestion: 
> 1. when Term(leader) < Term(follower), set resp.last_log_id to the index of first occurence of Term(follower) in follower's log entries. 
But once leader has receive the `resp.last_log_id` from follower, maybe the leader can't do anything, because the leader may lack some log in our implementation.
> when Term(leader) > Term(follower), just decrement last_log_id
This is more like the process in origin paper, the `decrement` will last for quite a long time to detect `nextIndex`.

Reply:
We could do it for further optimization, leave it alone in this PR.